core/src/main/scala/org/bykn/bosatsu/rankn/Infer.scala

```diff
@@ -448,7 +550,13 @@ object Infer {
         case (Type.TyApply(a1, b1), Type.TyApply(a2, b2)) =>
           unify(a1, a2, r1, r2) *> unify(b1, b2, r1, r2)
         case (Type.TyConst(c1), Type.TyConst(c2)) if c1 == c2 => unit
-        case (left, right) => fail(Error.NotUnifiable(left, right, r1, r2))
+        // these shouldn't be reachable since we have Tau types, but may be reachable
```

Comment:
this no longer fails with this PR, so the issue here is likely due to TyApply, and properly handling it fixes things.